Allow changing the author date from describe

OvidiusCicero opened the pull request:
This is unfinished as it needs tests, probably breaks tests and has room for improvement but I first want to evaluate whether this change has change to land before polishing. Also the CLI is up for debate (should this be its own command or is describe the right place?) (Also should similar nobs be introduced for committer information)

Proof it works
~~~
❯ jj show wpltwvzr
Commit ID: 10c39be28a519cfb067bc0fd616d06d091eb43c09d30dcb5fe138c2182fc6bfa2a99c0c105778ec317acb3007c3a9bfba6fa51fc6f4743acb08f20bdbe95f8e8
Change ID: wpltwvzrrymrzszypnnlywkyulqwtoyq
Author   : …  (1985-12-20 02:39:57)
Committer: … (2025-07-30 18:29:04)

    tests 2

Modified regular file test.txt:
   1    1: ugl<dugflguflzg
        2: lufgl<uaszgduoz
        3: iuglugl
        4: 
        5: öyfiuvhöydiuvg
        6: yfuvöiufgv
        7: iuguig
        8: 
        9: dsuvpuidgv

~/trasch/testj on ☁️  🥋 wplt | tests 2 
❯ ~/trasch/cargo/debug/jj desc -r wpltwvzr --author-date 1995-12-19T16:39:57-08:00 --no-edit
Working copy  (@) now at: wpltwvzr 2b8d46f1 tests 2
Parent commit (@-)      : rppnmuxk b7004367 test 1

~/trasch/testj on ☁️  🥋 wplt | tests 2 
❯ jj show wpltwvzr
Commit ID: 2b8d46f110a6fb71fc9788b8e166a3e38c2047fa696f60be9ddae6b1955449bbf023809281b9af81e7cc913209a2a9c349d5ebb211d6ab053c2fa4c9c41bc011
Change ID: wpltwvzrrymrzszypnnlywkyulqwtoyq
Author: … (1995-12-20 02:39:57)
Committer: … (2025-07-30 18:29:55)

    tests 2

Modified regular file test.txt:
   1    1: ugl<dugflguflzg
        2: lufgl<uaszgduoz
        3: iuglugl
        4: 
        5: öyfiuvhöydiuvg
        6: yfuvöiufgv
        7: iuguig
        8: 
        9: dsuvpuidgv
~~~


# Checklist

If applicable:

- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (`README.md`, `docs/`, `demos/`)
- [ ] I have updated the config schema (`cli/src/config-schema.json`)
- [x] I have added/updated tests to cover my changes

martinvonz commented:
What's the use case?

You can actually already do using this undocumented method:
```
JJ_TIMESTAMP=2000-01-01T00:00:00+00:00 jj describe --reset-author
```

I'm not sure if we should document that method or if we should add an argument like this PR does.

OvidiusCicero commented:
“undocumented”; maybe there was a misunderstanding but I did ask about it in the Discord before and the answer was JJ only supported this for name and E-Mail.

I have a slight preference for using flags instead of environment variables because then discovery is easier but I don't really care. It should be documented though, I need this functionality quite often tbh.

martinvonz commented:
We already have `--author`, so it seems consistent to add `--committer`, `--author-date`, and `--committer-date` too. What do you think?

> I need this functionality quite often tbh.

What do you use it for? Do you write tooling that imports commits from some other VCS?

joyously commented:
> I have a slight preference for using flags instead of environment variables

I can't tell which date will be set with `JJ_TIMESTAMP`. I also think a flag could be more forgiving about date format, and could use a function to get the date from some other commit.

> What do you use it for?

I've asked about it before because I have generated a repo from zip files (work from before I heard about VCS) and from someone else's project that wasn't in a VCS.

OvidiusCicero commented:
> We already have --author, so it seems consistent to add --committer, --author-date, and --committer-date too. What do you think?

yes, I'd add it

> What do you use it for? Do you write tooling that imports commits from some other VCS (that's what I've used the feature for in Git)?

Oh just something I edit via phone where I don't have access to command line and make commits (weeks) later

icorbrey commented:
> Why should people know or care when I first committed the first draft of a change? It's not even that useful since when you first committed something and when you first started working on it may be two completely different dates... 

I'd definitely like to echo this sentiment, I run into this all the time and Azure DevOps doesn't know how to handle it wrt PRs etc.

<img width="1080" height="253" alt="image" src="https://github.com/user-attachments/assets/89e0a328-347b-4207-b6d5-e3fc2cb3073a" />

AngelEzquerra commented:
> > > Why should people know or care when I first committed the first draft of a change? It's not even that useful since when you first committed something and when you first started working on it may be two completely different dates...
> > 
> > 
> > I'd definitely like to echo this sentiment, I run into this all the time and Azure DevOps doesn't know how to handle it wrt PRs etc.
> 
> FYI, the author date is set on `jj new` while the commiter date gets set whenever the commit is rewritten (the reason for this is #2000). I still like the older model though.

Thanks for the pointer to #2000. If I understand that discussion jj currently sets the author date when a commit becomes non discardable, right? I guess that's sensible _if you want to share the start of the work as well as the end of the work_ (which some people in the thread find useful).

IMHO, the author / committer dichotomy is yet another example of git overcomplicating things unnecessarily _and_ exposing that complexity to the user by default. Pretty much every other VCS I know doesn't have that concept and they get by fine without it.

I think it'd be better to hide the author / committer dichotomy by default. In the same way that jj has simplified git's workflow in many ways, it could do the same here are remove yet another layer of git's complexity (or at least hide it by default).

OvidiusCicero commented:
Added tests, documented the commands and added the same functionality with respect to committer. There is one TODO I would need your oppinion on it's reviewable from my stand

(sorry I was distracted the last weeks)

OvidiusCicero commented:
uhm can someone explain to me why the test fails in CI but not locally for me?

glehmann commented:
I successfully ran the test suite with your branch on my computer. Not sure what is happening…

OvidiusCicero commented:
it might have something to do with the date parsing now being dependent on `now()` and us both @glehmann sitting in the same time zone (central europe). Will have a look in the next days. If someone can enlighten me I'd still be grateful

yuja commented:
Appears that `interim` converts timezone internally to the `now` timezone, so there's no way to know the timezone specified by the user?

https://github.com/conradludgate/interim?tab=readme-ov-file#api

OvidiusCicero commented:
and log output does not show time zones, so you only see it in a different commit ID even though the test explicitly sets a time zone.

@yuja what follows from this? Revert back to the “home grown” combination of two std functions and move this to `lib/time_utils.rs`? Still use this but fix the test, knowing this breaks tests locally and yields wrong results even when inputting a time zone explicitly? (not liking the latter one)

Fixing interim is probably out of reach for me

OvidiusCicero commented on lib/src/time_util.rs:
```diff
@@ -122,6 +122,16 @@ impl DatePattern {
+/// Parse a string with time zone information into a `Timestamp`
+pub fn parse_datetime(s: &str) -> chrono::ParseResult<Timestamp> {
+    Ok(Timestamp::from_datetime(
+        DateTime::parse_from_rfc2822(s).or_else(|_| DateTime::parse_from_rfc3339(s))?,
```
added a unit test and a system test

OvidiusCicero commented on cli/tests/test_touch_command.rs:
```diff
@@ -240,32 +240,142 @@ fn test_touch() {
+
+    // explicitly test full time stamp to check for time zone
+    insta::assert_snapshot!(work_dir.run_jj(["log", "-r", "@", "-T", "author.timestamp()"]).success(), @r"
+    @  1995-12-19 16:39:57.000 -08:00
```
done

OvidiusCicero commented on cli/src/commands/touch.rs:
```diff
@@ -81,6 +83,24 @@ pub(crate) struct TouchArgs {
+        long,
+        value_parser = parse_author
+    )]
+    committer: Option<(String, String)>,
```
actually lets leave it out
